working/macros/feature-specification.md

```diff
@@ -319,6 +319,63 @@ be constructed, but where you need to build up something in smaller pieces you
 can use the [Fragment][] subtype. Any arbitrary String can be passed to this
 class, allowing you to build up your code fragments however you like.
 
+### Adding New Macro Applications
+
+Macros are allowed to add new macro applications in two ways:
+
+#### Adding Macro Applications to New Declarations
+
+When creating [Code][] instances, a macro may generate code which includes
+macro applications. These macro applications must be from either the current
+phase or a later phase, but cannot be from previous phases.
+
+If a macro application is added which implements an earlier phase, that phase
+is not ran. This should result in a warning if the macro does not also
+implement some phase that will be ran.
+
+If a macro application is added which runs in the same phase as the current
+one, then it is immediately expanded after execution of the current macro,
+following the normal ordering rules.
+
+#### Adding Macro Applications to Existing Declarations
+
+Macro applications can be added to existing declarations through the `*Builder`
+APIs. Macros added in this way are always prepended to the list of existing
+macros on the declaration (which makes them run _last_).
+
+Note that macros can already _immediately_ invoke another macro on a given
+declaration manually, by simply instantiating the macro and then invoking
+it.
+
+TODO: Update the builder apis to allow this.
+
+#### Note About Ordering Violations
+
+Note that both of these mechanisms allow for normal macro ordering to be
+circumvented. Consider the following example, where all macros run in the
+Declaration phase:
+
+```dart
+@macroA
+@macroB
+class X {
+  @macroC // Added by `@macroA`, runs after both `@macroB` and `@macroA`
+  int? a; 
+
+  // Generated by `@macroC`, not visible to `@macroB`.
+  int? b;
+}
+```
+
+Normally, macros always run "inside-out". But in this case `@macroC` runs after
+both `@macroB` and `@macroC` which were applied to the class.
+
+We still allow this because it doesn't cause any ambiguity in ordering, even
+though it violates the normal rules.
+
+We could instead only allow adding macros from _later_ phases, but that would
+restrict the functionality in unnecessary ways.
```

Comment:
Do we have known use cases for macros generating code that contains macro applications? Are there cases where the macro can't simply directly invoke the corresponding macro code to achieve the same effect?

Unless we know this is needed, I'd be very inclined to leave it out because, if nothing else, it's pretty confusing for users to reason about.

Reply:
Yes we have examples - namely this comes up when you want to hook into lifecycle methods whether the user implements them already or not. If the method exists you want to apply yourself to it, and if it doesn't you want to generate a version of it with yourself applied.